common/scala/src/main/scala/whisk/common/tracing/TraceUtil.scala

```diff
@@ -0,0 +1,141 @@
+    * @param transactionId transactionId to which this Trace belongs.
+    * @return TracedRequest which provides details about current service being traced.
+    */
+  def startTrace(serviceName: String, spanName: String, transactionId: TransactionId): Unit = {
+    traceMap.get(transactionId.meta.id) match {
+      case Some(tracedRequest) => {
+        startChildTrace(serviceName, spanName, tracedRequest, transactionId)
+      }
+      case None => {
+        var tracedRequest = createTracedRequest(spanName)
+        var metadata = trace.sample(tracedRequest.request, serviceName)
+        if (metadata != None) {
+          trace.start(tracedRequest.request, serviceName)
+          tracedRequest = new TracedRequest(tracedRequest.request, null, metadata)
+          traceMap.put(transactionId.meta.id, tracedRequest)
+        }
+      }
+    }
+
+  }
+
+  /**
+    * Start a Child Trace
+    * @param serviceName Name of service to be traced
+    * @param parent Parent TracedRequest to continue span propagation and create span hierarchy.
+    * @param transactionId transactionId to which this Trace belongs.
+    * @return TracedRequest which provides details about current service being traced.
+    */
+  def startChildTrace(serviceName: String, spanName: String, parent: TracedRequest, transactionId: TransactionId): Unit = {
+    val tracedReq: TracedRequest = createTracedRequest(spanName)
+    var metadata =  trace.sample(tracedReq.request, serviceName)
+    if(metadata.isDefined){
+      trace.createChild(tracedReq.request, parent.request)
+      trace.start(tracedReq.request, serviceName)
+      val tracedRequest = new TracedRequest(tracedReq.request, Some(parent), metadata)
+      traceMap.put(transactionId.meta.id, tracedRequest)
+    }
+  }
+
+  /**
+    * Finish a Trace associated with given transactionId.
+    * @param transactionId
+    */
+  def finish(transactionId: TransactionId): Unit = {
+    traceMap.get(transactionId.meta.id) match {
+      case Some(tracedRequest) => {
+        trace.record(tracedRequest.request, TracingAnnotations.ServerSend)
+        traceMap.remove(transactionId.meta.id)
+        if(tracedRequest.parent.isDefined)
+          traceMap.put(transactionId.meta.id, tracedRequest.parent.get)
+      }
+      case _ =>
+    }
+  }
+
+  def error(transactionId: TransactionId, t: Throwable) : Unit = {
+    val tracedRequest = traceMap.get(transactionId.meta.id)
+    if(tracedRequest.isDefined){
+      trace.record(tracedRequest.get.request, t)
+    }
+  }
+
+  private def createTracedRequest(spanName: String, metadata: Option[SpanMetadata] = None) : TracedRequest = {
+    val headers: Map[String, Integer] = Map[String, Integer]()
+    headers.put("id", requestCounter.getAndIncrement())
+    val req: BasicTraceRequest = new BasicTraceRequest(headers, spanName)
+
+    if(metadata.isDefined)
+      trace.importMetadata(req, metadata.get, spanName)
+
+    return TracedRequest(req, null, metadata)
```

Comment:
In Scala one does not use return. While this will work it might have unwanted side-effects in certain situations. I'd recommend to not use return at all in Scala.

Reference: https://tpolecat.github.io/2014/05/09/return.html